promql/analyzer.go

```diff
@@ -0,0 +1,155 @@
+	// the range into the past is given by the duration, as in "foo[5m]".
+	ranges map[clientmodel.Fingerprint]time.Duration
+}
+
+// analyze the provided expression and attach metrics and fingerprints to data-selecting
+// AST nodes that are later used to preload the data from the storage.
+func (a *Analyzer) analyze(ctx context.Context) {
+	a.offsetPreloadTimes = map[time.Duration]preloadTimes{}
+
+	getPreloadTimes := func(offset time.Duration) preloadTimes {
+		if _, ok := a.offsetPreloadTimes[offset]; !ok {
+			a.offsetPreloadTimes[offset] = preloadTimes{
+				instants: map[clientmodel.Fingerprint]struct{}{},
+				ranges:   map[clientmodel.Fingerprint]time.Duration{},
+			}
+		}
+		return a.offsetPreloadTimes[offset]
+	}
+
+	// Retrieve fingerprints and metrics for the required time range for
+	// each metric or matrix selector node.
+	Inspect(a.Expr, func(node Node) bool {
+		switch n := node.(type) {
+		case *VectorSelector:
+			pt := getPreloadTimes(n.Offset)
+			fpts := a.ng.storage.GetFingerprintsForLabelMatchers(n.LabelMatchers)
+			n.fingerprints = fpts
+			n.metrics = map[clientmodel.Fingerprint]clientmodel.COWMetric{}
+			n.iterators = map[clientmodel.Fingerprint]local.SeriesIterator{}
+			for _, fp := range fpts {
+				// Only add the fingerprint to the instants if not yet present in the
+				// ranges. Ranges always contain more points and span more time than
+				// instants for the same offset.
+				if _, alreadyInRanges := pt.ranges[fp]; !alreadyInRanges {
+					pt.instants[fp] = struct{}{}
+				}
+				n.metrics[fp] = a.ng.storage.GetMetricForFingerprint(fp)
+			}
+		case *MatrixSelector:
+			pt := getPreloadTimes(n.Offset)
+			fpts := a.ng.storage.GetFingerprintsForLabelMatchers(n.LabelMatchers)
+			n.fingerprints = fpts
+			n.metrics = map[clientmodel.Fingerprint]clientmodel.COWMetric{}
+			n.iterators = map[clientmodel.Fingerprint]local.SeriesIterator{}
+			for _, fp := range fpts {
+				if pt.ranges[fp] < n.Range {
+					pt.ranges[fp] = n.Range
+					// Delete the fingerprint from the instants. Ranges always contain more
+					// points and span more time than instants, so we don't need to track
+					// an instant for the same fingerprint, should we have one.
+					delete(pt.instants, fp)
+				}
+				n.metrics[fp] = a.ng.storage.GetMetricForFingerprint(fp)
+			}
+		}
+		return true
+	})
+}
+
+// prepare the expression evaluation by preloading all required chunks from the storage
+// and setting the respective storage iterators in the AST nodes.
+func (a *Analyzer) prepare(ctx context.Context) (local.Preloader, error) {
+	if a.offsetPreloadTimes == nil {
+		return nil, errors.New("analysis must be performed before preparing query")
+	}
+	// The preloader must not be closed unless an error ocurred as closing
+	// unpins the preloaded chunks.
+	p := a.ng.storage.NewPreloader()
+
+	// Preload all analyzed ranges.
+	for offset, pt := range a.offsetPreloadTimes {
+		a.ng.checkContext(ctx)
```

Comment:
Usage of the `context` package had in mind that the code base might grow, query handling may spread out into different packages, etc. Thus, the function should probably be exported.
In general we might only export stuff that's actually used in other places, but aside from an application we are also creating packages. Their design should enable future use in previously unintended ways (c.f. yesterday's PR that exported the `prepare*` functions).

I'd propose to return `error` rather than `bool`. Callers don't have to create repetitive error messages this way (and check if it was cancellation or timeout for that). It also allows us to mask the default errors from `context` with something more user friendly.

```
if err := promql.ContextDone(ctx, "query analysis"); err != nil {
    return nil, err
}
```
As this is asking for cycles, this function might be better off in a utility package.